Documentation/botr/clr-abi.md

```diff
@@ -0,0 +1,550 @@
+
+## Allocating local variables backward
+
+This is required to preserve addresses of the existing locals when an EnC edit appends new ones. In other words, the first local must be allocated at the highest stack address. Special care has to be taken to deal with alignment. The total size of the method frame can either grow (more locals added) or shrink (fewer temps needed) after the edit. The VM zeros out newly added locals.
+
+## Fixed set of callee-saved registers
+
+This eliminates need to deal with the different sets in the VM, and makes preservation of local addresses easier. On x64, we choose to always save `RBP` only. There are plenty of volatile registers and so lack of non-volatile registers does not impact quality of non-optimized code.
+
+## EnC is supported for methods with EH
+
+However, EnC remap is not supported inside funclets. The stack layout of funclets does not matter for EnC.
+
+## Initial RSP == RBP == PSPSym
+
+This invariant allows VM to compute new value of `RBP` and PSPSym after the edit without any additional information. Location of PSPSym is found via GC info.
+
+## Localloc
+
+Localloc is allowed in EnC code, but remap is disallowed after the method has executed a localloc instruction. VM uses the invariant above (`RSP == RBP`) to detect whether localloc was executed by the method.
+
+## Security object
+
+This does not require any special handling by the JIT on x64. (Different from x86). The security object is copied over by the VM during remap if necessary. Location of security object is found via GC info.
+
+## Synchronized methods
+
+The extra state created by the JIT for synchronized methods (original "this" and lock taken flag) must be preserved during remap. The JIT stores this state in the preserved region, and increases the size of the preserved region reported in GC info accordingly.
+
+## Generics
+
+EnC is not supported for generic methods and methods on generic types.
+
+# System V x86_64 support
+
+This section relates mostly to calling conventions on System V systems (such as Ubuntu Linux and Mac OS X).
+The general rules outlined in the System V x86_64 ABI (described at http://www.x86-64.org/documentation/abi.pdf) are followed with a few exceptions, described below:
+
+1. The hidden argument for by-value passed structs is always after the "this" parameter (if there is one). This is a difference with the System V ABI and affects only the internal JIT calling conventions. For PInvoke calls the hidden argument is always the first parameter since there is no "this" parameter in this case.
+2. Managed structs that have no fields are always passed by-value on the stack.
+3. The JIT proactively generates frame register frames (with `RBP` as a frame register) in order to aid the native OS tooling for stack unwinding and the like.
+4. All the other internal VM contracts for PInvoke, EH, and generic support remains in place. Please see the relevant sections above for more details.
+
```

Comment:
One more thing that comes to mind is that structs with explicit layouts are always passed by value on the stack.